spec-main/api-session-spec.js

```diff
@@ -0,0 +1,559 @@
+      expect(list[0]).to.have.property('name', name)
+      expect(list[0]).to.have.property('value', value)
+      expect(list[0]).to.have.property('domain', domain)
+    })
+
+    it('emits a changed event when a cookie is added or removed', async () => {
+      const changes = []
+
+      const { cookies } = session.fromPartition('cookies-changed')
+      const name = 'foo'
+      const value = 'bar'
+      const listener = (event, cookie, cause, removed) => { changes.push({ cookie, cause, removed }) }
+
+      const a = emittedOnce(cookies, 'changed')
+      await cookies.set({ url, name, value, expirationDate: (+new Date()) / 1000 + 120 })
+      const [, setEventCookie, setEventCause, setEventRemoved] = await a
+
+      const b = emittedOnce(cookies, 'changed')
+      await cookies.remove(url, name)
+      const [, removeEventCookie, removeEventCause, removeEventRemoved] = await b
+
+      expect(setEventCookie.name).to.equal(name)
+      expect(setEventCookie.value).to.equal(value)
+      expect(setEventCause).to.equal('explicit')
+      expect(setEventRemoved).to.equal(false)
+
+      expect(removeEventCookie.name).to.equal(name)
+      expect(removeEventCookie.value).to.equal(value)
+      expect(removeEventCause).to.equal('explicit')
+      expect(removeEventRemoved).to.equal(true)
+    })
+
+    describe('ses.cookies.flushStore()', async () => {
+      describe('flushes the cookies to disk', async () => {
+        const name = 'foo'
+        const value = 'bar'
+        const { cookies } = session.defaultSession
+
+        await cookies.set({ url, name, value })
+        await cookies.flushStore()
+      })
+    })
+
+    it('should survive an app restart for persistent partition', async () => {
+      const appPath = path.join(fixtures, 'api', 'cookie-app')
+      const electronPath = process.execPath
+
+      const test = (result, phase) => {
+        return new Promise((resolve, reject) => {
+          let output = ''
+
+          const appProcess = ChildProcess.spawn(
+            electronPath,
+            [appPath],
+            { env: { PHASE: phase, ...process.env } }
+          )
+
+          appProcess.stdout.on('data', data => { output += data })
+          appProcess.stdout.on('end', () => {
+            output = output.replace(/(\r\n|\n|\r)/gm, '')
+            expect(output).to.equal(result)
+            resolve()
+          })
+        })
+      }
+
+      await test('011', 'one')
+      await test('110', 'two')
+    })
+  })
+
+  describe('ses.clearStorageData(options)', () => {
+    it('clears localstorage data', async () => {
+      const { ipcMain } = require('electron')
+      await w.loadFile(path.join(fixtures, 'api', 'localstorage.html'))
+      const options = {
+        origin: 'file://',
+        storages: ['localstorage'],
+        quotas: ['persistent']
+      }
+      await w.webContents.session.clearStorageData(options)
+      while (await w.webContents.executeJavaScript('localStorage.length') !== 0) {
```

Comment:
expose internal storage api in the browser process for testing purpose.

Reply:
i'm not sure. this api will be changing as part of nws13n so i'm inclined to leave it this way for now.